docker/docker-entrypoint.sh

```diff
@@ -16,6 +16,17 @@ web)
   shift
   set -- gunicorn "promgen.wsgi:application" "$@"
   ;;
+wait-runserver)
```

Comment:
Instead of using a separate wait-runserver block, I wonder if this would be better added to the existing 'web' block above here. Runserver is sufficient for development but perhaps the docker container should use gunicorn.

Reply:
I realized that my Dockerfile entrypoint uses *web* though my original docker-compose uses *runserver*. It's probably best to make both as *web* since *runserver* is mostly for development use. What do you think?

I also thought a bit about having the `promgen check ; sleep` check apply to all of the commands by putting it at the top, but there are also likely valid cases or not forcing that check while debugging the system. Do you have any thoughts about this as well?

Reply:
gunicorn is running a WSGI server, so this would require an additional container running nginx (or similar) to be able to serve HTTP requests.

I think it is sensible to have two different entrypoints, both with a wait mechanism:

 * An entrypoint starting up a HTTP server using the runserver command .
 * An entrypoint starting up a WSGI server using gunicorn.

I think commands should be regarded individually when considering whether to wait for db. With one-off commands run manually, it is in the user's interest to get an immediate response. For example `worker` does not really need to wait for db, as it's basic functionality does not depend on the availability of mysql and resilience to redis outages is already built into celery.

Perhaps for some one-off commands it would be useful to have optional wait for db functionality, for example when running it from a docker-compose file. In this case either a new command `wait-…` could be introduced using case fallthrough or an env var could be checked to decide whether to wait for db or try immediately.

Reply:
gunicorn can speak HTTP so nginx is not required though it is often used in a production environment. I do not have a strong opinion since I tend to consider docker-compose a bit more of a demo than a fully ready production environment. When deploying for real it's likely that most would want to tweak their setup a bit.

Maybe best to keep things simple in this pass. Can always add additional wait- commands or make it global if required in the future.

Reply:
Hm, what do you think about these entrypoints, both protected by `promgen check`:

 * `web` - gunicorn running the application. Used as the default for docker-compose.
 * `web-dev` - django `runserver`. Meant to be used for testing and development.

`web` could then be used for a more production-y setup together with nginx. Additional documentation could be compiled on how to achieve this (e.g. using two containers sharing a volume for hosting of the static files).

Reply:
Sounds reasonable. I think those names are more descriptive than *wait-runserver* for those who may not be as familiar with django directly.